### ISSUE TITLE
Segfaults pretty much instantly using NS

### ISSUE THREAD

**Issue by casv2** [ACEsuit/ACEinterfaces#13]
ACEinterfaces segfaults after ~50 steps using NS. Pyjulip also segfaults occasionally but only every ~100K steps or so. I've gone back to using PyJulip for NS now...

**Comment by davkovacs** [comment]
I am aware of the segfaults after 100k-200k steps, and have no idea what the cause might be. I will do my best to figure out asap!

**Comment by cortner** [comment]
@gabor1 whats the chance of hiring an RSE for this? None of us are really competent here. We could even get a quote from Julia computing. Probably in the range 3-10k. What do you think?

**Comment by gabor1** [comment]
Let's get a quote. We are depending on this daily

**Comment by gabor1** [comment]
The whole python<-julia needs a robust and sustainable solution

**Comment by gabor1** [comment]
It's much easier now that you have an example where it happens right away

**Comment by casv2** [comment]
Segfaults like these used to happen pretty instantly using this MWE for pyjulia I posted two years ago: https://github.com/JuliaPy/pyjulia/issues/306. Since then a lot has changed, Python, Julia, PyJulia everything...

I rewrote the MWE for the "new" PyJulia, and ran with latest Julia v1.7 and Python 3.9.5
```julia
module LJ
using LinearAlgebra

function energy(R::Matrix{T}) where {T}
   E = zero(T)
   for i = 1:size(R, 2)-1, j = i+1:size(R,2)
      r= norm(R[:,i] - R[:,j])
      E += r^(-12) - 2 * r^(-6)
   end
   return E
end
end
```
and
```python
from julia.api import Julia
jl = Julia(compiled_modules=False)
import numpy as np

from julia import Main
Main.eval('include(\"LJ.jl\")')

for j in range(100000):
    e = Main.LJ.energy(np.random.rand(10,3))
    if j % 50 == 0:
        print(j,e)
```

and it actually ran fine... I also updated pyjulip today and am running it using NS, ACE1, Julia v1.7 and Python 3.9.5 in anger the next few days to check whether I can get it to segfault. So far so good...

Also, isn't pyjulia built on this `julia_init()` C-interface business? Aren't we effectively going through the same thing through ACEinterfaces? Is there a bug deep inside Julia and how it communicates with any other language?

But yes I agree we need something really really robust...

**Comment by cortner** [comment]
Yes we are using the same concepts but our thought was that we don’t want to start maintaining pyjulia but only a very very small subset.

**Comment by cortner** [comment]
@davkovacs Should we have a pyjulia backend for your interface code? Then we can compare directly.

**Comment by casv2** [comment]
I understand and agree it makes much more sense to maintain something we control, but what if the concepts you mention we rely on are buggy? e.g. something deep inside Julia itself

**Comment by cortner** [comment]
I actually doubt that the Julia embedding per se is buggy.

**Comment by davkovacs** [comment]
The reason why it would be important to have the C -> Julia interface robust, is because it is very versatile. Not just for the ASE Python calculator, but I reused large parts of it for the OpenMM C++ code to ACE.jl interface as well. And I see the same problem of seg fault after 250k steps there. 
Having said that I am not opposed to having a PyJulia version for the ACEInterfaces ASE python calculator. But that will not quite solve all the problems.

**Comment by casv2** [comment]
Ofcourse, we need this to work. And checking this morning one pyjulip NS run failed, so I restarted... So pyjulip/pyjulia still isn't bullit proof either... Took 10K NS iterations (keep in mind that NS is parallelised over 32 cores and does like ~1K force evalations for an 'iteration' per core, my rough estimate is that this is equivalent to tens of millions of MD steps on a molecule, but who cares really, it's still buggy).

**Comment by gabor1** [comment]
But doesn't this suggest that the problem is not with the python but deeper? what is the common element of the python and c interfaces?

**Comment by casv2** [comment]
That's what I was thinking... Julia embedding or something really really deep like that. My understanding is that pyjulia and ACE interfaces are basically built on top of this Julia embedding.

Perhaps we're just the first people to really stress test this part of Julia? Is it possible that no one else has done millions/billions of calls through this embedding? Not sure...

**Comment by davkovacs** [comment]
I will try to create a C++ example that I can run in valgrind next week. It would be good if that spotted something. (though I have never used valgrind, so I am not quite sure what I am looking for) 
I will try to work with the main OpenMM developer guy (he doesn't know Julia, but is a great software engineer and might be able to help. They seem quite keen to have as many different ML models in OpenMM as possible. )

**Comment by casv2** [comment]
Regarding that Pyjulip segfault I mentioned earlier, I'm pretty sure that was just a hole in the ACE model itself... Been running several NS jobs using a much better ACE for a few days now with no issues. Configurations go crazy rather quickly near a hole in the potential, in turn leading to configs failing to meet asserts in the ACE code. This leads to Julia terminating which I think could be causing (some) of our segfaults?

I tried out ACEinterfaces using this ACE model I trust quite well, but that failed again really quite quickly. But as I said, using pyjulip it's been running fine for days. It's probably good to realise that occasionally these segfaults could be due to the ACE potential being holey.

**Comment by gabor1** [comment]
why does a hole give a segfault? it should be caught, right?

**Comment by casv2** [comment]
In HAL, yes, in NS, no. We can try adding this feature though, we'd need to put it in the ASEcalculator but first implement uncertainty estimates in ACE1 and warn when we're starting to get near a hole in NS. Perhaps even trigger a DFT, but while you're still in the gas phase that's a bit crazy. Also uncertainty estimates in gas would probably all over the shop? So probably unusable anyways?

**Comment by bernstei** [comment]
NS (pymatnest at least) should catch python exception, but I _have_ seen lots of instances of julia errors leading to seg faults rather than python exceptions.

**Comment by casv2** [comment]
Indeed, and it's impossible to say whether it was the Julia->C or Julia->Python connection or the ACE model itself failing, right? I've looked at the segfaults occasionally and I can't tell at least.

**Comment by cortner** [comment]
Can you give me examples of those asserts you are talking about?

**Comment by bernstei** [comment]
When I was having the ACE -> ACE1 transition problems, e.g. the old json files having identifiers like "ACE_" instead of "ACE1_" (eventually fixed with the sed replace one liner), I believe that those led to seg faults, not nicely catchable exceptions, when called via ACEInterfaces.

**Comment by cortner** [comment]
Thanks - so any exception leads to a segfaukt? Strange. 

So we need to identify the exceptions that might be triggered in the evaluation stage now. What about rij = 0?

**Comment by bernstei** [comment]
One type of seg fault happens when you try to use a previous version json (tags with `ACE_` instead of `ACE1_`), in the call to the actual energy calculation:

```
tin 1010 : python t.py 

signal (11): Segmentation fault
in expression starting at none:0
typekeyvalue_hash at /buildworker/worker/package_linux64/build/src/jltypes.c:1152 [inlined]
lookup_typevalue at /buildworker/worker/package_linux64/build/src/jltypes.c:722
jl_inst_arg_tuple_type at /buildworker/worker/package_linux64/build/src/jltypes.c:1589
arg_type_tuple at /buildworker/worker/package_linux64/build/src/gf.c:1845 [inlined]
jl_lookup_generic_ at /buildworker/worker/package_linux64/build/src/gf.c:2373 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2425
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1788 [inlined]
jl_call2 at /buildworker/worker/package_linux64/build/src/jlapi.c:256
energy at /home/cluster2/bernstei/src/work/ACE/ACEinterfaces/ase/ace/ace_c.c:92
ffi_call_unix64 at /home/Software/python/anaconda3/lib/python3.8/lib-dynload/../../libffi.so.8 (unknown line)
ffi_call_int at /home/Software/python/anaconda3/lib/python3.8/lib-dynload/../../libffi.so.8 (unknown line)
_call_function_pointer at /usr/local/src/conda/python-3.8.12/Modules/_ctypes/callproc.c:921 [inlined]
_ctypes_callproc at /usr/local/src/conda/python-3.8.12/Modules/_ctypes/callproc.c:1264
PyCFuncPtr_call at /usr/local/src/conda/python-3.8.12/Modules/_ctypes/_ctypes.c:4201
_PyObject_MakeTpCall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyFunction_Vectorcall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyFunction_Vectorcall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyEval_EvalCodeWithName at python (unknown line)
_PyFunction_Vectorcall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyEval_EvalCodeWithName at python (unknown line)
_PyFunction_Vectorcall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyEval_EvalCodeWithName at python (unknown line)
_PyFunction_Vectorcall at python (unknown line)
_PyEval_EvalFrameDefault at python (unknown line)
_PyEval_EvalCodeWithName at python (unknown line)
PyEval_EvalCodeEx at python (unknown line)
PyEval_EvalCode at python (unknown line)
unknown function (ip: 0x5634318f62e2)
unknown function (ip: 0x563431912542)
unknown function (ip: 0x563431917561)
PyRun_SimpleFileExFlags at python (unknown line)
Py_RunMain at python (unknown line)
Py_BytesMain at python (unknown line)
__libc_start_main at /lib64/libc.so.6 (unknown line)
unknown function (ip: 0x56343188ad68)
Allocations: 21049366 (Pool: 21044806; Big: 4560); GC: 17
Segmentation fault (core dumped)
```

**Comment by casv2** [comment]
When you evaluate an ACE1 on a particularly silly structure (atoms literally all over the place) it complains about some assert on a cosine theta somewhere. Might even be in JuLIP.jl or NeighbourLists.jl? I've seen it when you're very very far away from anything even remotely sensible. It crashes Julia and probably also raises a segfault I think. 

Do you want me to try and reproduce? Maybe we can implement some error handling of this particular error?

**Comment by casv2** [comment]
After some digging I found it: `@assert abs(S.cosθ) <= 1.0` in `/src/polynomials/sphericalharmonics.jl` (line 303) is sometimes violated when you're deep inside a hole

**Comment by cortner** [comment]
This is helpful - likely soherucal harmonics evaluation

**Comment by cortner** [comment]
Can you create a reproducible test pkease

**Comment by cortner** [comment]
Sorry I didn’t see your previous message . Thanks for tracking it down, that’s exactly what I was wondering. Please create a new branch to debug this, then convert the assert to an if statement and print out detailed information. After that DONT fail but project the cos to [-1, 1] . If it’s and NaN then just fail with an error call

**Comment by cortner** [comment]
If you can print out the full vector at t which the assert fails?

**Comment by casv2** [comment]
Thanks Christoph, I'll try out what you mentioned above.

Just for completeness, NS failed after many days of running using PyJulip. Here's the message
```
free(): invalid pointer

signal (6): Aborted
in expression starting at none:0
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
abort at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x15125bf75966)
unknown function (ip: 0x15125bf7c9d9)
cfree at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
_buffer_info_free at /home/casv2/miniconda3/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so (unknown line)
_dealloc_cached_buffer_info at /home/casv2/miniconda3/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so (unknown line)
array_dealloc at /home/casv2/miniconda3/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so (unknown line)
array_assign_subscript at /home/casv2/miniconda3/lib/python3.8/site-packages/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so (unknown line)
_PyEval_EvalFrameDefault at python3 (unknown line)
_PyFunction_Vectorcall at python3 (unknown line)
unknown function (ip: 0x561246219b83)
_PyFunction_Vectorcall at python3 (unknown line)
unknown function (ip: 0x561246219b83)
_PyEval_EvalCodeWithName at python3 (unknown line)
_PyFunction_Vectorcall at python3 (unknown line)
unknown function (ip: 0x561246219b83)
_PyFunction_Vectorcall at python3 (unknown line)
unknown function (ip: 0x56124621b77e)
_PyEval_EvalCodeWithName at python3 (unknown line)
PyEval_EvalCodeEx at python3 (unknown line)
PyEval_EvalCode at python3 (unknown line)
unknown function (ip: 0x561246334f83)
unknown function (ip: 0x5612463671f3)
PyRun_FileExFlags at python3 (unknown line)
PyRun_SimpleFileExFlags at python3 (unknown line)
unknown function (ip: 0x56124623098a)
Py_BytesMain at python3 (unknown line)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 0x5612462f9e92)
```

Does say something about a pointer... No mention of a segfault in the message explicitly?

**Comment by gelzinyte** [comment]
Hi, I thought I'd make a note of my experiences with regards to segmentation faults. 

In this particular case I was running 200 000 step MD on `segy` branch of ACE1 and `debug` branch of @jameskermode's ACEInterfaces fork. The MDs failed 9 times out of 30. Updates in `debug` helped a lot - before switching only 3/150 trajectories completed successfully... 

It doesn't seem to be a potential issue - I could restart the trajectories a step or two before the segfault, they ran through the point of failure and the structures were entirely reasonable. Also all 15 MDs, restarted from the same initial structures/momenta as above, I tried to run finished successfully with pyjulip, which is what I'm using now.

**Comment by bernstei** [comment]
Were your restarts _identical_, or just from configurations saved with "reasonable" precision?  Because those won't be exactly identical.

**Comment by gelzinyte** [comment]
No, I didn't try to make them _identical_ and trajectories aren't identical, but I thought I'd highlight that the MDs that I'm comparing aren't very different.

**Comment by davkovacs** [comment]
I am very happy that the changes we made helped somewhat. But it is still very annoying, and it would require a more robust solution. Sorry about this Elena...
